src/ir/predicates/predicate_test.cc

```diff
@@ -0,0 +1,88 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "src/ir/predicates/predicate.h"
+
+#include "absl/strings/str_replace.h"
+#include "absl/strings/substitute.h"
+#include "google/protobuf/text_format.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/predicates/predicate_textproto_to_rule_body_testdata.h"
+#include "src/ir/proto/predicate.h"
+#include "third_party/arcs/proto/manifest.pb.h"
+
+namespace raksha::ir::predicates {
+
+namespace proto = raksha::ir::proto;
+
+class ToDatalogRuleBodyTest :
+ public testing::TestWithParam<
+  std::tuple<std::tuple<absl::string_view, absl::string_view>, AccessPath>> {
+ public:
+  ToDatalogRuleBodyTest() :
+    textproto_(std::get<0>(std::get<0>(GetParam()))),
+    access_path_(std::get<1>(GetParam())) {
+    absl::string_view rule_body_sub_string =
+        std::get<1>(std::get<0>(GetParam()));
+    expected_rule_body_ =
+        absl::Substitute(rule_body_sub_string, access_path_.ToString());
+  }
```

Comment:
nit: any way to make this more readable?

Reply:
I'll see what I can do. Since I've basically been forced to write a constructor body anyway, I might do some more of this work there rather than in the per-field initializers.

Reply:
I think I improved it by moving most stuff out of the field initializers. I had to leave the one for `access_path_` there because it has no default constructor. I think it's a pretty big improvement though.